Fixes #36084 - set config.action_mailer.raise_delivery_errors = true

Ron-Lavi opened the pull request:
in case that delivery_method is "sendmail",
we should also check if the executable file exist. currently the mail can be seen as sent succesfully from the UI, although the executable to send it is missing


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

Ron-Lavi commented:
to make it work, and to have an existing sendmail_location,
you can follow the following installation instructions - e.g on fedora: https://tecadmin.net/install-sendmail-on-fedora/

ekohl commented on app/controllers/users_controller.rb:
```diff
@@ -193,6 +193,11 @@ def test_mail
         return
       end
+
+      if (Setting[:delivery_method].to_s == "sendmail" && !File.exist?(Setting[:sendmail_location]))
```
Your solution does check if the file exists, but doesn't check if it's executable or some permission error (SELinux could deny access for example).

https://guides.rubyonrails.org/action_mailer_basics.html#example-action-mailer-configuration suggests there's an option `config.action_mailer.raise_delivery_errors = true` and I think that's what we're actually after here. It may also catch errors on the SMTP delivery so I'd appreciate it if you investigate that option.

ekohl reviewed (changes requested):
> you can follow the following installation instructions - e.g on fedora: https://tecadmin.net/install-sendmail-on-fedora/

Note that sendmail is not something you really want. It may have been the only choice when it was written 40 years ago (yes, it's that old), but nowadays `/usr/bin/sendmail` is more of an interface. For example, postfix also provides it. On my Fedora:
```console
# dnf repoquery --whatprovides /usr/sbin/sendmail -q
esmtp-0:1.2-19.fc36.x86_64
exim-0:4.95-3.fc36.x86_64
exim-0:4.96-6.fc36.x86_64
opensmtpd-0:6.8.0p2-10.fc36.x86_64
opensmtpd-0:6.8.0p2-8.fc36.x86_64
postfix-2:3.6.4-1.fc36.x86_64
sendmail-0:8.17.1-5.fc36.x86_64
ssmtp-0:2.64-29.fc36.x86_64
```

Red Hat defaults to postfix because it's much more user friendly than sendmail.

ronlavi2412 committed:
Fixes #36084 - set `config.action_mailer.raise_delivery_errors = true`

when this setting was disabled, the test email always showed success,
even when the mail test actually failed.

ekohl reviewed (commented):
I don't know if we always need to error out. That implies we may need to catch it where we call it and display an error. I was mostly talking about the test mail.

@adamruzicka did you also mention this was better in the latest mailer gem?

Ron-Lavi commented:
> It may have been the only choice when it was written 40 years ago (yes, it's that old), 

Well, seemed that it served well for so many years :)

>but nowadays /usr/bin/sendmail is more of an interface. For example, postfix also provides it. On my Fedora:
> `dnf repoquery --whatprovides /usr/sbin/sendmail -q`

Nice, `--whatprovides` command is super useful! TIL

@ekohl @adamruzicka looks like the `config.action_mailer.raise_delivery_errors = true` solved the issue.

stejskalleos commented:
[test katello]

stejskalleos commented:
:green_apple: **LGTM** Error is generated as expected.

Let's wait for Katello results, meanwhile @ekohl do you any other comments? If not we can merge

ekohl commented:
> Let's wait for Katello results, meanwhile @ekohl do you any other comments? If not we can merge

I still think that this is missing error handling. My biggest worry is that if an email is sent, but failed to deliver that an exception is raised. Then the controller which sends the email may display an error page (HTTP 500) while the primary action worked.

One way around that is to start using [`deliver_later`](https://api.rubyonrails.org/v7.0.4.2/classes/ActionMailer/MessageDelivery.html#method-i-deliver_later) (https://projects.theforeman.org/issues/36118).

I didn't test this, but I suppose one way to verify it would be:
* Configure audit entries to be mailed
* Make sure email can't be delivered (like setting sendmail to a location that doesn't exist)
* Edit a page that creates an audit entry

My theory would be that it shows an internal server error page.

Another would be user creation if welcome emails are enabled.

The `test_email` does have exception handling, so that should work just fine.

Ron-Lavi commented:
Alright then, I will update this PR with deliver_later.

ronlavi2412 committed:
Refs #36084 - replace deliver_now with deliver_later

ronlavi2412 committed:
Refs #36084 - Add initializer to deal with mails exceptions
